### PR TITLE
Issue 863  finite set

### PR DESCRIPTION
**Pull request by adler-j**
Adds three rather basic sets

### CODE REVIEW

File: odl/set/sets.py [hunk 1]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+            has the same subsets as this set, `False` otherwise.
+        """
+        return (type(self) == type(other) and
+                len(other) == len(self) and
+                all(so == ss for so, ss in zip(other.sets, self.sets)))
+
+    def element(self, inp=None):
+        """Create a new element.
+
+        First tries calling the first set, then the second, etc.
```
Comment (adler-j) [inline odl/set/sets.py]: update

File: odl/set/sets.py [hunk 2]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+    def __len__(self):
+        """Return ``len(self)``."""
+        return len(self.sets)
+
+    def __getitem__(self, indcs):
+        """Return ``self[indcs]``.
+
+        Examples
+        --------
+        >>> emp, univ = odl.EmptySet(), odl.UniversalSet()
```
Comment (adler-j) [inline odl/set/sets.py]: Perhaps use some other example sets

File: odl/set/sets.py [hunk 3]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+            has the same subsets as this set, `False` otherwise.
+        """
+        return (type(self) == type(other) and
+                len(other) == len(self) and
+                all(eo == es for eo, es in zip(other.elements, self.elements)))
+
+    def element(self, inp=None):
+        """Create a new element.
+
+        First tries calling the first set, then the second, etc.
```
Comment (adler-j) [inline odl/set/sets.py]: update

File: odl/set/sets.py [hunk 4]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+            return inp
+        else:
+            return Set.element(self, inp)
+
+    def __getitem__(self, indcs):
+        """Return ``self[indcs]``.
+
+        Examples
+        --------
+
```
Comment (adler-j) [inline odl/set/sets.py]: Remove blank row

**Comment by adler-j** [comment]
bump

File: odl/set/sets.py [hunk 5]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+        Parameters
+        ----------
+        set1, ..., setN : `Set`
+            The elements in the set.
+
+        Examples
+        --------
+        >>> set = odl.FiniteSet(1, 'string')
+        """
+        self.__elements = tuple(elements)
```
Comment (kohr-h) [inline odl/set/sets.py]: How about duplicates?

File: odl/set/sets.py [hunk 6]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+
+        First tries calling the first set, then the second, etc.
+
+        For more specific control, use set[i].element() to pick which subset to
+        use.
+        """
+        if inp in self.elements:
+            return inp
+        else:
+            return Set.element(self, inp)
```
Comment (kohr-h) [inline odl/set/sets.py]: Huh? Does this raise an error or what?

File: odl/set/sets.py [hunk 7]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+
+        Returns
+        -------
+        equals : `boolean`
+            `True` if `other` is a `SetUnion` instance, and
+            has the same subsets as this set, `False` otherwise.
+        """
+        return (type(self) == type(other) and
+                len(other) == len(self) and
+                all(eo == es for eo, es in zip(other.elements, self.elements)))
```
Comment (kohr-h) [inline odl/set/sets.py]: `other.elements == self.elements`??

File: odl/set/sets.py [hunk 8]
```diff
@@ -621,6 +622,352 @@ def __repr__(self):
...
+        if not all(isinstance(set_, Set) for set_ in sets):
+            wrong = [set_ for set_ in sets
+                     if not isinstance(set_, Set)]
+            raise TypeError('{!r} not Set instance(s)'.format(wrong))
+
+        self.__sets = tuple(sets)
+
+    @property
+    def sets(self):
+        """The factors (sets) as a tuple."""
```
Comment (adler-j) [inline odl/set/sets.py]: Any suggestion?

**Comment by adler-j** [comment]
Thanks for the review, I'll got most of it fixed asap. Anyway regarding the wider comments, yes this is largely a "because we can", but also it is so fundamental to mathematics that I guess our users would assume that we support it.

The reason this is made instead of the built-in `set` and `tuple` is largely that we need a thin wrapper in order to use those with methods that only work on sets (see e.g. an operator with output `True` or `False`, could be given range `FiniteSet(True, False)`.)

File: odl/set/sets.py (hunk 8 above)
Comment (kohr-h) [inline odl/set/sets.py]:
```python
docstrings = ["The sets of this {} as a tuple.".format(x)
              for x in ('union', 'intersection')]
```
😉

**Comment by kohr-h** [comment]
> Thanks for the review, I'll got most of it fixed asap. Anyway regarding the wider comments, yes this is largely a "because we can", but also it is so fundamental to mathematics that I guess our users would assume that we support it.

Well, there's a lot more fundamental math to cover. But in a sense, because we already have abstract `Set`s, adding a simple special case and some basic operations on it is not totally far-fetched.

I'm trying hard to think of a case where *within the ODL ecosystem* you would need these classes. Functions to finite sets are not very useful (in principle yes, but not with this generality) because vectorization is impossible, or at least checks like `contains_all` become quite expensive.
Functionals... I don't know.
Operators... maybe


> The reason this is made instead of the built-in set and tuple is largely that we need a thin wrapper in order to use those with methods that only work on sets (see e.g. an operator with output True or False, could be given range FiniteSet(True, False).)

That was my suspicion 😉 An example doing this would certainly be more convincing.

File: odl/set/sets.py (hunk 5 above)
Comment (adler-j) [inline odl/set/sets.py]: Also `__eq__` is order dependent :/

**Comment by adler-j** [comment]
Fixed comments, need another review.

**Comment by adler-j** [comment]
Bump

File: odl/util/utility.py [hunk 9]
```diff
@@ -23,15 +23,17 @@
 standard_library.install_aliases()
 
 from functools import wraps
+from collections import OrderedDict
 import numpy as np
 
 
+
```
Comment (kohr-h) [inline odl/util/utility.py]: remove extra line

File: odl/set/sets.py [hunk 10]
```diff
@@ -521,7 +521,7 @@ def __init__(self, *sets):
 
     @property
     def sets(self):
-        """Factors (sets) as a tuple."""
+        """The sets of this carthesian product as a tuple."""
```
Comment (kohr-h) [inline odl/set/sets.py]: wrong spelling

File: odl/set/sets.py [hunk 11]
```diff
@@ -642,31 +637,32 @@ def __init__(self, *sets):
         ----------
         set1, ..., setN : `Set`
             The sets whose union should be taken.
+            Any duplicates are be ignored.
```
Comment (kohr-h) [inline odl/set/sets.py]: ~~be~~

File: odl/set/sets.py [hunk 12]
```diff
@@ -684,13 +680,17 @@ def __eq__(self, other):
...
         """
         return (type(self) == type(other) and
-                len(other) == len(self) and
-                all(so == ss for so, ss in zip(other.sets, self.sets)))
+                all(set_ in other for set_ in self) and
+                all(set_ in self for set_ in other))
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), set(self.sets)))
```
Comment (kohr-h) [inline odl/set/sets.py]: Why convert to `set` before hashing?

File: odl/set/sets.py [hunk 13]
```diff
@@ -717,12 +717,12 @@ def __getitem__(self, indcs):
...
         Examples
         --------
-        >>> emp, univ = odl.EmptySet(), odl.UniversalSet()
-        >>> prod = odl.SetUnion(emp, univ, univ, emp, emp)
-        >>> prod[2]
-        UniversalSet()
-        >>> prod[2:4]
-        SetUnion(UniversalSet(), EmptySet())
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> prod = odl.SetUnion(reals, complexnrs)
```
Comment (kohr-h) [inline odl/set/sets.py]: `prod` not a good name

File: odl/set/sets.py [hunk 14]
```diff
@@ -762,31 +758,32 @@ def __init__(self, *sets):
         ----------
         set1, ..., setN : `Set`
             The sets whose intersection should be taken.
+            Any duplicates are be ignored.
```
Comment (kohr-h) [inline odl/set/sets.py]: ~~be~~

File: odl/set/sets.py [hunk 15]
```diff
@@ -837,12 +822,12 @@ def __getitem__(self, indcs):
...
         Examples
         --------
-        >>> emp, univ = odl.EmptySet(), odl.UniversalSet()
-        >>> prod = odl.SetIntersection(emp, univ, univ, emp, emp)
-        >>> prod[2]
-        UniversalSet()
-        >>> prod[2:4]
-        SetIntersection(UniversalSet(), EmptySet())
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> prod = odl.SetIntersection(reals, complexnrs)
```
Comment (kohr-h) [inline odl/set/sets.py]: `prod` not good

File: odl/set/sets.py [hunk 16]
```diff
@@ -918,33 +895,38 @@ def __eq__(self, other):
 
         Returns
         -------
-        equals : `boolean`
-            `True` if `other` is a `SetUnion` instance, and
-            has the same subsets as this set, `False` otherwise.
+        equals : bool
+            ``True`` if ``other`` is a `SetUnion` instance, and
```
Comment (kohr-h) [inline odl/set/sets.py]: update

File: odl/util/utility.py [hunk 17]
```diff
@@ -707,6 +709,44 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+    Parameters
+    ----------
+    seq : sequence
+        Sequence with (possibly duplicate) elements.
+
+    Returns
+    -------
+    unique : list
+        Unique elements of seq. Order is guaranteed to be the same as in seq.
```
Comment (kohr-h) [inline odl/util/utility.py]: \`\`seq\`\`

File: odl/util/utility.py [hunk 18]
```diff
@@ -707,6 +709,44 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+
+    Examples
+    --------
+    >>> unique([1, 2, 3, 3])
+    [1, 2, 3]
+
+    >>> unique((1, 'str', 'str'))
+    [1, 'str']
+
+    Also works with unhashable types:
```
Comment (kohr-h) [inline odl/util/utility.py]: Make a sentence that doesn't start with "Also".

File: odl/set/sets.py (hunk 1 above)
Comment (kohr-h) [inline odl/set/sets.py]: Done?

File: odl/set/sets.py [hunk 19]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+
+        For more specific control, use set[i].element() to pick which subset to
+        use.
+        """
+        for set in self.sets:
+            try:
+                return set.element(inp)
+            except NotImplementedError:
+                pass
+        raise NotImplementedError('element not implemented for any of the '
```
Comment (kohr-h) [inline odl/set/sets.py]: \`element\`

File: odl/set/sets.py [hunk 20]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), set(self.sets)))
+
+    def element(self, inp=None):
+        """Create a new element.
+
+        First tries calling the first set, then the second, etc.
+
+        For more specific control, use set[i].element() to pick which subset to
```
Comment (kohr-h) [inline odl/set/sets.py]: typeset `set[i].element()`

File: odl/set/sets.py [hunk 21]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+        """Return ``len(self)``."""
+        return len(self.sets)
+
+    def __getitem__(self, indcs):
+        """Return ``self[indcs]``.
+
+        Examples
+        --------
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> prod = odl.SetUnion(reals, complexnrs)
```
Comment (kohr-h) [inline odl/set/sets.py]: union

File: odl/set/sets.py [hunk 22]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
         return '{}({})'.format(self.__class__.__name__, sets_str)
 
 
+class SetUnion(Set):
+    """The union of several subsets.
```
Comment (kohr-h) [inline odl/set/sets.py]: blank line above

File: odl/set/sets.py [hunk 23]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> odl.SetUnion(reals, complexnrs)
+        SetUnion(RealNumbers(), ComplexNumbers())
+        """
+        sets_str = ', '.join(repr(set_) for set_ in self.sets)
+        return '{}({})'.format(self.__class__.__name__, sets_str)
+
+
+class SetIntersection(Set):
+    """The intersection of several subsets.
```
Comment (kohr-h) [inline odl/set/sets.py]: blank above

Check with the docstring of `CartesianProduct`.

File: odl/set/sets.py [hunk 24]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+        """Return ``len(self)``."""
+        return len(self.sets)
+
+    def __getitem__(self, indcs):
+        """Return ``self[indcs]``.
+
+        Examples
+        --------
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> prod = odl.SetIntersection(reals, complexnrs)
```
Comment (kohr-h) [inline odl/set/sets.py]: intersection

File: odl/set/sets.py [hunk 25]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+        >>> reals, complexnrs = odl.RealNumbers(), odl.ComplexNumbers()
+        >>> odl.SetIntersection(reals, complexnrs)
+        SetIntersection(RealNumbers(), ComplexNumbers())
+        """
+        sets_str = ', '.join(repr(set_) for set_ in self.sets)
+        return '{}({})'.format(self.__class__.__name__, sets_str)
+
+
+class FiniteSet(Set):
+    """A set given by a finite number of elements."""
```
Comment (kohr-h) [inline odl/set/sets.py]: blank above

File: odl/set/sets.py [hunk 26]
```diff
@@ -621,6 +621,335 @@ def __repr__(self):
...
+            has the same subsets as this set, ``False`` otherwise.
+        """
+        # Need to loop since order could be different
+        return (type(self) == type(other) and
+                all(el in other for el in self) and
+                all(el in self for el in other))
+
+    def __hash__(self):
+        """Return ``hash(self)``."""
+        return hash((type(self), set(self.elements)))
```
Comment (kohr-h) [inline odl/set/sets.py]: Why go through `set`?

File: odl/util/utility.py [hunk 27]
```diff
@@ -707,6 +709,44 @@ def signature_string(posargs, optargs, sep=', ', mod=''):
...
+
+    >>> unique((1, [1], [1]))
+    [1, [1]]
+    """
+    # First check if all elements are hashable, if so O(n) can be done
+    try:
+        return list(OrderedDict.fromkeys(seq))
```
Comment (kohr-h) [inline odl/util/utility.py]: Is there no `OrderedSet` or similar?

**Comment by kohr-h** [comment]
Status on this? Test failure is only a PEP8 thing.

**Comment by adler-j** [comment]
I'll try to get this out of the way quite soon, but as you note there are quite some work.

File: odl/util/utility.py (hunk 27 above)
Comment (adler-j) [inline odl/util/utility.py]: It seems not:
https://stackoverflow.com/questions/1653970/does-python-have-an-ordered-set

File: odl/set/sets.py (hunk 12 above)
Comment (adler-j) [inline odl/set/sets.py]: To ignore ordering

File: odl/set/sets.py (hunk 26 above)
Comment (adler-j) [inline odl/set/sets.py]: Uniqueness again

**Pull request merged**